concept/manager/ConceptListenerImpl.java

```diff
@@ -135,20 +136,19 @@ private void thingCreated(Thing thing, boolean isInferred) {
     }
 
     @Override
-    public <D> void attributeCreated(Attribute<D> attribute, D value, boolean isInferred) {
+    public <D> void attributeCreated(Attribute<D> attribute, String uniqueIndex, boolean isInferred) {
         Type type = attribute.type();
         //Track the attribute by index
         Label label = type.label();
-        String index = Schema.generateAttributeIndex(label, value.toString());
-        transactionCache.addNewAttribute(label, index, attribute.id());
+        transactionCache.addNewAttribute(label, uniqueIndex, attribute.id());
         thingCreated(attribute, isInferred);
-        attributeManager.ackAttributeInsert(index, txId);
+        attributeManager.ackAttributeInsert(uniqueIndex, txId);
     }
 
     private <D> void attributeDeleted(Attribute<D> attribute) {
-        Type type = attribute.type();
+        AttributeType<D> type = attribute.type();
         //Track the attribute by index
-        String index = Schema.generateAttributeIndex(type.label(), attribute.value().toString());
+        String index = Schema.generateAttributeIndex(type.label(), AttributeValueConverter.tryConvert(type, attribute.value()).toString());
```

Comment:
again, I don't think we need to apply the (potentially costly? not sure!) conversion everywhere - only where the user provides a value, it comes from a query, or a rule, etc! So I think this and the following cases can be kept as before as the value comes from a pre-existing `Attribute` instance :)

Reply:
I'm not so convinced about this one. That `Attribute` could have come from many places - going up the stack, we have:

- ConceptNotificationChannel.hasAttributeCreated
- ConceptManager.createHasAttribute
- ThingImpl.attributeOwnership
  - Thing.has
    - HasAttributeExecutor.InsertHasAttribute ...
    - ConceptMethod.run
  - Thing.attributeInferred
    - AttributeMaterialiser.putAttributeOwmership
    - AttributeMaterialiser.materialise
    - AttributeAtom ...